### PR TITLE
Improvements to the network topologies doc

### PR DESCRIPTION
**Pull request by rvennam**
Improvements to https://preliminary.istio.io/latest/docs/ops/configuration/traffic-management/network-topologies/

[ ] Configuration Infrastructure
[X ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: content/en/docs/ops/configuration/traffic-management/network-topologies/index.md [hunk 1]
```diff
@@ -24,56 +24,34 @@ You might choose to deploy Istio ingress gateways in various network topologies
 Istio ingress gateway to the Internet). As such, these topologies require different ingress gateway configurations for
 transporting correct client attributes like IP addresses and certificates to the workloads running in the cluster.
 
-Configuration of XFF and XFCC headers is managed via `MeshConfig` during Istio
-installation or by adding a pod annotation. Note that the `Meshconfig` configuration is a global setting for all gateway workloads, while pod annotations override the global setting on a per-workload basis.
+Configuration of XFF and XFCC headers can be set globally for all gateway workloads via `MeshConfig` or per gateway using a pod annotation. For example, to configure globally during install when using an `IstioOpeartor` CR:
```
Comment (nrjpoddar) [inline content/en/docs/ops/configuration/traffic-management/network-topologies/index.md:27]:
```suggestion
Configuration of XFF and XFCC headers can be set globally for all gateway workloads via `MeshConfig` or per gateway using a pod annotation. For example, to configure globally during install or upgrade when using an `IstioOpeartor` CR:
```

File: content/en/docs/ops/configuration/traffic-management/network-topologies/index.md [hunk 2]
```diff
@@ -100,10 +78,6 @@ to understand how `X-Forwarded-For` headers and trusted client addresses are det
     $ istioctl install -f topology.yaml
     {{< /text >}}
 
-    {{< idea >}}
```
Comment (nrjpoddar) [inline content/en/docs/ops/configuration/traffic-management/network-topologies/index.md:103]: Restarting proxies is still needed IIRC. Are you seeing that the Operator does a rolling upgrade for gateways?

**Review by nrjpoddar — COMMENTED** [review]
I like it! Should we add a warning in the beginning about this doesn’t support extracting client attributes via Proxy protocol and you need EnvoyFilter resource to insert the proxy protocol filter.

File: content/en/docs/ops/configuration/traffic-management/network-topologies/index.md (hunk 2 above)
Comment (linsun) [reply content/en/docs/ops/configuration/traffic-management/network-topologies/index.md:103]: Yes i had to restart pod afterwards also

**Comment by linsun** [comment]
> I like it! Should we add a warning in the beginning about this doesn’t support extracting client attributes via Proxy protocol and you need EnvoyFilter resource to insert the proxy protocol filter.


+1 good clarification from @rvennam on this page!  So I ran the step yesterday and found proxy protocol is also part of the headers for httpbin, in step 7:

```
$ curl -H 'X-Forwarded-For: 56.5.6.7, 72.9.5.6, 98.1.2.3, 1.1.1.1'  http://$GATEWAY_URL/get\?show_env\=true
{
  "args": {
    "show_env": "true"
  }, 
  "headers": {
    "Accept": "*/*", 
    "Content-Length": "0", 
    "Host": "169.60.80.54", 
    "User-Agent": "curl/7.64.1", 
    "X-B3-Parentspanid": "5a4c6502198946a1", 
    "X-B3-Sampled": "0", 
    "X-B3-Spanid": "bafa7d6e0f667dca", 
    "X-B3-Traceid": "d30aa17ac46832145a4c6502198946a1", 
    "X-Envoy-Attempt-Count": "1", 
    "X-Envoy-External-Address": "98.1.2.3", 
    "X-Forwarded-Client-Cert": "By=spiffe://cluster.local/ns/default/sa/httpbin;Hash=c327e632625c2805f2c488fc81a735a28e1c1a9d053a1d50f395932d070aa73c;Subject=\"\";URI=spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account", 
    "X-Forwarded-For": "56.5.6.7, 72.9.5.6, 98.1.2.3, 1.1.1.1,10.188.142.194", 
    "X-Forwarded-Proto": "http", 
    "X-Request-Id": "65b101c5-f231-4553-b35a-a6e8aca9a64a"
  }, 
  "origin": "56.5.6.7, 72.9.5.6, 98.1.2.3, 1.1.1.1,10.188.142.194", 
  "url": "http://169.60.80.54/get?show_env=true"
}
```
I think our doc should add `X-Forwarded-Proto` in the output as well and explain this doc page can be also used for XFP.  @nrjpoddar I would prefer not use envoy filter if this serves users' need for proxy protocol.

cc @howardjohn for input as well.

File: content/en/docs/ops/configuration/traffic-management/network-topologies/index.md [hunk 3]
```diff
@@ -24,56 +24,34 @@ You might choose to deploy Istio ingress gateways in various network topologies
 Istio ingress gateway to the Internet). As such, these topologies require different ingress gateway configurations for
 transporting correct client attributes like IP addresses and certificates to the workloads running in the cluster.
 
-Configuration of XFF and XFCC headers is managed via `MeshConfig` during Istio
-installation or by adding a pod annotation. Note that the `Meshconfig` configuration is a global setting for all gateway workloads, while pod annotations override the global setting on a per-workload basis.
+Configuration of XFF and XFCC headers can be set globally for all gateway workloads via `MeshConfig` or per gateway using a pod annotation. For example, to configure globally during install or upgrade when using an `IstioOpeartor` CR:
```
Comment (nrjpoddar) [inline content/en/docs/ops/configuration/traffic-management/network-topologies/index.md:27]:
```suggestion
Configuration of XFF and XFCC headers can be set globally for all gateway workloads via `MeshConfig` or per gateway using a pod annotation. For example, to configure globally during install or upgrade when using an `IstioOperator` custom resource:
```

File: content/en/docs/ops/configuration/traffic-management/network-topologies/index.md [hunk 4]
```diff
@@ -155,9 +129,8 @@ to understand how `X-Forwarded-For` headers and trusted client addresses are det
     }
     {{< /text >}}
 
-Note that the `X-Envoy-External-Address` is set to the "second" from last address in the `X-Forwarded-For` header
-as per your `numTrustedProxies` setting. Additionally, the gateway workload appends its IP in the
-`X-Forwarded-For` header before forwarding it to the upstream httpbin workload.
+The above output shows the headers that the httpbin workload received. When the Istio gateway received this request, it set the `X-Envoy-External-Address` header to the second to last (`numTrustedProxies: 2`) address in the `X-Forwarded-For` header from your curl command. Additionally, the gateway appends its own IP to the
```
Comment (nrjpoddar) [inline content/en/docs/ops/configuration/traffic-management/network-topologies/index.md:132]:
```suggestion
The above output shows the request headers that the `httpbin` workload received. When the Istio gateway received this request, it set the `X-Envoy-External-Address` header to the second to last (`numTrustedProxies: 2`) address in the `X-Forwarded-For` header from your curl command. Additionally, the gateway appends its own IP to the
```

**Comment by nrjpoddar** [comment]
@linsun I'm not following your comment. If you remove the XFF header from the client request, I'm assuming Gateway proxy will not automatically parse the proxy protocol connection string without configuring an EnvoyFilter. Is that still not the case?

**Comment by linsun** [comment]
@nrjpoddar @rvennam playing a bit more with this with master code and here is my understanding of things:

- the numTrustedProxies pod annotation only helps user to calculate the `X-Envoy-External-Address`. For example, if i'm not configuring numTrustedProxies, by default, my `X-Envoy-External-Address` is the last IP from `X-Forwarded-For`

```
    "X-Envoy-External-Address": "192.168.2.71", 
    "X-Forwarded-Client-Cert": "By=spiffe://cluster.local/ns/default/sa/httpbin;Hash=3c7ae71d64ef532686c4ef50ab8a71a629915b0b283196996e3eacf696995b80;Subject=\"\";URI=spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account", 
    "X-Forwarded-For": "56.5.6.7, 72.9.5.6, 98.1.2.3,192.168.2.71",
```
If I had the numTrustedProxies annotation on my pod, it would help me calculate the correct value for `X-Envoy-External-Address` based on value of numTrustedProxies.

The proxy protocol can be tested using `curl url --haproxy-protocol`. The gateway network topology config doesn't help with this at all. I had it confused with `X-Forwarded-Proto` earlier.  If user needs proxy protocol enabled on the gateway, users will need to apply a proxy protocol filter to the listener port that they want proxy protocol enabled.

Examples I tested:

```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: proxy-protocol
  namespace: istio-system
spec:
  configPatches:
  - applyTo: LISTENER
    match:
      listener:
        portNumber: 8080
    patch:
      operation: MERGE
      value:
        listener_filters:
        - name: envoy.listener.proxy_protocol
        - name: envoy.listener.tls_inspector
  workloadSelector:
    labels:
      istio: ingressgateway
```

```
$ curl $GATEWAY_URL/get\?show_env\=true --haproxy-protocol               
{
  "args": {
    "show_env": "true"
  }, 
  "headers": {
    "Accept": "*/*", 
    "Content-Length": "0", 
    "Host": "169.60.80.54", 
    "User-Agent": "curl/7.64.1", 
    "X-B3-Parentspanid": "31ce8b465f47c5d4", 
    "X-B3-Sampled": "0", 
    "X-B3-Spanid": "e78fa538f639c648", 
    "X-B3-Traceid": "e80a0c485649a98231ce8b465f47c5d4", 
    "X-Envoy-Attempt-Count": "1", 
    "X-Envoy-Internal": "true", 
    "X-Forwarded-Client-Cert": "By=spiffe://cluster.local/ns/default/sa/httpbin;Hash=3c7ae71d64ef532686c4ef50ab8a71a629915b0b283196996e3eacf696995b80;Subject=\"\";URI=spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account", 
    "X-Forwarded-For": "192.168.2.71", 
    "X-Forwarded-Proto": "http", 
    "X-Request-Id": "bdabfba6-09e3-45eb-a8cc-cb4aa2761669"
  }, 
  "origin": "192.168.2.71", 
  "url": "http://169.60.80.54/get?show_env=true"
}
```

When proxy protocol filter is enabled, from the [doc](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/listener_filters/proxy_protocol),  the Proxy Protocol must be present on the connection (either version 1 or version 2), the standard does not allow parsing to determine if it is present or not.  This is why this now fails:

```
$ curl $GATEWAY_URL/get\?show_env\=true
curl: (56) Recv failure: Connection reset by peer
```

So best is to enable this per port number as needed.  For example my 15017 port isn't impacted.

Thoughts going forward - should we add proxyProtocol as a field in `gatewayTopology` to allow users to easily enable this per a given port?

**Comment by rvennam** [comment]
Thanks for investigating and clarifying everything @linsun! 

Agree that we need this API support for proxy protocol. There are many users asking for this and EnvoyFilter is the only solution right now. 

@nrjpoddar , I will add a bit about proxy protocol to this doc.

**Comment by nrjpoddar** [comment]
@linsun your explanation is in-line with how this should work. There was a lot of controversy around API for proxy protocol so I left it out of my initial RFC. Even getting API ratified took me > 2 months..

File: content/en/docs/ops/configuration/traffic-management/network-topologies/index.md (hunk 2 above)
Comment (rvennam) [reply content/en/docs/ops/configuration/traffic-management/network-topologies/index.md]: Added it back in the example task section

**Comment by rvennam** [comment]
- Added a section for PROXY protocol
- Tweaked title `Gateway Proxy Network Topology (Development)`
- Added a sentence about `X-Envoy-External-Address` header

File: content/en/docs/ops/configuration/traffic-management/network-topologies/index.md [hunk 5]
```diff
@@ -24,56 +24,34 @@ You might choose to deploy Istio ingress gateways in various network topologies
...
+      "proxy.istio.io/config": '{"gatewayTopology" : { "numTrustedProxies": <VALUE>, "forwardClientCertDetails": <ENUM_VALUE> } }'
 {{< /text >}}
 
 ### Configuring X-Forwarded-For Headers
 
 Applications rely on reverse proxies to forward client attributes in a request, such as `X-Forward-For` header. However, due to the variety of network
-topologies Istio can be deployed in, you must set the number of trusted proxies deployed in front
-of the Istio gateway proxy, so that the client address can be extracted correctly.
+topologies that Istio can be deployed in, you must set the `numTrustedProxies` to the number of trusted proxies deployed in front
+of the Istio gateway proxy, so that the client address can be extracted correctly. This helps the Istio Gateway set the correct value for the `X-Envoy-External-Address` header.
```
Comment (nrjpoddar) [inline content/en/docs/ops/configuration/traffic-management/network-topologies/index.md:52]:
```suggestion
of the Istio gateway proxy, so that the client address can be extracted correctly.
This controls the value populated by the ingress gateway in the `X-Envoy-External-Address` header
which can be reliably used by the upstream services to access client's original IP address.
```
```suggestion
of the Istio gateway proxy, so that the client address can be extracted correctly. This helps the Istio Gateway set the correct value for the `X-Envoy-External-Address` header.
```

File: content/en/docs/ops/configuration/traffic-management/network-topologies/index.md [hunk 6]
```diff
@@ -195,3 +172,27 @@ where `ENUM_VALUE` can be of the following type.
 
 See the [Envoy documentation](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-forwarded-client-cert)
 for examples of using this capability.
+
+## PROXY Protocol
+
+The [PROXY protocol](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) allows for exchanging and preservation of client attributes across multiple proxies. Enabling this requires adding [Envoy Proxy Protocol filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/listener_filters/proxy_protocol), using an `EnvoyFilter` applied on the gateway workload. For example:
```
Comment (nrjpoddar) [inline content/en/docs/ops/configuration/traffic-management/network-topologies/index.md:178]:
```suggestion
The [PROXY protocol](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) allows for exchanging and preservation of client attributes across multiple proxies without relying on Layer 7 protocols.
Enabling this requires adding [Envoy Proxy Protocol filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/listener_filters/proxy_protocol), using an `EnvoyFilter` applied on the gateway workload. For example:
```
```suggestion
The [PROXY protocol](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) allows for exchanging and preservation of client attributes across multiple proxies. Enabling this requires adding [Envoy Proxy Protocol filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/listener_filters/proxy_protocol), using an `EnvoyFilter` applied on the gateway workload. For example:
```

**Review by nrjpoddar — COMMENTED** [review]
Adding the Proxy protocol config is good but that section needs beefing up.

1. What happens when you enable PROXY protocol listener and numTrustedProxies is set and a client sends XFF headers. Which setting takes precedence?
2. When PROXY protocol is enabled, can a malicious client fake add CONNECTION string and will Envoy populate the external address?

**Comment by linsun** [comment]
We had a discussion about this topic in networking WG.  @nrjpoddar filled us with background info and why this API is not appropriate for mesh config.  Post 1.8, we should try find a good API home for user to enable this. 

@nrjpoddar also asked about precedence of proxy protocol vs gateway topology and here is data I collected:

1. curl with --haproxy-protocol set with no XFF headers in request. Gateway listener filter set and no topology configuration. What's the request  `X-Envoy-External-Address` header received at the upstream service after gateway proxies it.
$ curl $GATEWAY_URL/get\?show_env\=true --haproxy-protocol
{
  "args": {
    "show_env": "true"
  }, 
  "headers": {
    "Accept": "*/*", 
    "Content-Length": "0", 
    "Host": "169.60.80.54", 
    "User-Agent": "curl/7.64.1", 
    "X-B3-Parentspanid": "8dda4a00e4cb39b7", 
    "X-B3-Sampled": "0", 
    "X-B3-Spanid": "470005758c7eb6a5", 
    "X-B3-Traceid": "668f7038f5d1801a8dda4a00e4cb39b7", 
    "X-Envoy-Attempt-Count": "1", 
    "X-Envoy-Internal": "true", 
    "X-Forwarded-Client-Cert": "By=spiffe://cluster.local/ns/default/sa/httpbin;Hash=e365cd8aba3b1042d0914678ca37bf8b52c1fc00d234ad686309a20ec1e436df;Subject=\"\";URI=spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account", 
    "X-Forwarded-For": "192.168.2.71", 
    "X-Forwarded-Proto": "http", 
    "X-Request-Id": "f5f23b02-17ac-49cf-9441-b999d3a1a399"
  }, 
  "origin": "192.168.2.71", 
  "url": "http://169.60.80.54/get?show_env=true"
}
2. curl with --haproxy-protocol set with  XFF headers in request. Gateway listener filter set and no topology configuration. What's the request  `X-Envoy-External-Address` header received at the upstream service after gateway proxies it.
$ curl -H 'X-Forwarded-For: 56.5.6.7, 72.9.5.6, 98.1.2.3' $GATEWAY_URL/get\?show_env\=true  --haproxy-protocol
{
  "args": {
    "show_env": "true"
  }, 
  "headers": {
    "Accept": "*/*", 
    "Content-Length": "0", 
    "Host": "169.60.80.54", 
    "User-Agent": "curl/7.64.1", 
    "X-B3-Parentspanid": "d5e3ddd9648767f6", 
    "X-B3-Sampled": "0", 
    "X-B3-Spanid": "ff8f13f600f784be", 
    "X-B3-Traceid": "6c471929db5bb27fd5e3ddd9648767f6", 
    "X-Envoy-Attempt-Count": "1", 
    "X-Envoy-External-Address": "192.168.2.71", 
    "X-Forwarded-Client-Cert": "By=spiffe://cluster.local/ns/default/sa/httpbin;Hash=e365cd8aba3b1042d0914678ca37bf8b52c1fc00d234ad686309a20ec1e436df;Subject=\"\";URI=spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account", 
    "X-Forwarded-For": "56.5.6.7, 72.9.5.6, 98.1.2.3,192.168.2.71", 
    "X-Forwarded-Proto": "http", 
    "X-Request-Id": "52c1ef31-ad7c-491f-8a58-20126613ea13"
  }, 
  "origin": "56.5.6.7, 72.9.5.6, 98.1.2.3,192.168.2.71", 
  "url": "http://169.60.80.54/get?show_env=true"
}
3. curl with --haproxy-protocol set with  XFF headers in request. Gateway listener filter set and  topology configuration. What's the request  `X-Envoy-External-Address` header received at the upstream service after gateway proxies it.
$ curl -H 'X-Forwarded-For: 56.5.6.7, 72.9.5.6, 98.1.2.3' $GATEWAY_URL/get\?show_env\=true  --haproxy-protocol
{
  "args": {
    "show_env": "true"
  }, 
  "headers": {
    "Accept": "*/*", 
    "Content-Length": "0", 
    "Host": "169.60.80.54", 
    "User-Agent": "curl/7.64.1", 
    "X-B3-Parentspanid": "d31a766ec8239855", 
    "X-B3-Sampled": "0", 
    "X-B3-Spanid": "830376c9e149a19f", 
    "X-B3-Traceid": "9d5ce505a1a1b194d31a766ec8239855", 
    "X-Envoy-Attempt-Count": "1", 
    "X-Envoy-External-Address": "72.9.5.6", 
    "X-Forwarded-Client-Cert": "By=spiffe://cluster.local/ns/default/sa/httpbin;Hash=43e95171af735dc119217b042ef02cabc8ffd352b144ab80adbe1578cc25c7ca;Subject=\"\";URI=spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account", 
    "X-Forwarded-For": "56.5.6.7, 72.9.5.6, 98.1.2.3,192.168.2.71", 
    "X-Forwarded-Proto": "http", 
    "X-Request-Id": "3d9649e0-96d9-4426-8b70-98379a740a9c"
  }, 
  "origin": "56.5.6.7, 72.9.5.6, 98.1.2.3,192.168.2.71", 
  "url": "http://169.60.80.54/get?show_env=true"
}

Conclusion: gateway topology annotation of numTrustedProxies take precedence while proxy protocol is enabled.

cc @howardjohn @rvennam

**Comment by linsun** [comment]
Another experiment: with proxy protocol filter + gateway topology numTrustedProxies enabled, without XFF header:

```
$ curl $GATEWAY_URL/get\?show_env\=true  --haproxy-protocol 
{
  "args": {
    "show_env": "true"
  }, 
  "headers": {
    "Accept": "*/*", 
    "Content-Length": "0", 
    "Host": "169.60.80.54", 
    "User-Agent": "curl/7.64.1", 
    "X-B3-Parentspanid": "572ec1a6466d3881", 
    "X-B3-Sampled": "0", 
    "X-B3-Spanid": "8cff59320eb33be6", 
    "X-B3-Traceid": "badd7003d6f298e1572ec1a6466d3881", 
    "X-Envoy-Attempt-Count": "1", 
    "X-Envoy-Internal": "true", 
    "X-Forwarded-Client-Cert": "By=spiffe://cluster.local/ns/default/sa/httpbin;Hash=ac5c88cfcb78967b0dbc466145be6dcf3d97505b6457b0bed3339359664b3a6b;Subject=\"\";URI=spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account", 
    "X-Forwarded-For": "192.168.2.71", 
    "X-Forwarded-Proto": "http", 
    "X-Request-Id": "0f36cc63-3d64-49da-9381-0b9544f90e7a"
  }, 
  "origin": "192.168.2.71", 
  "url": "http://169.60.80.54/get?show_env=true"
}
```

Note: when XFF header is not specified in client call, no IP is specified for `X-Envoy-External-Address`.  The `X-Forwarded-For` IP is correct, which is my laptop's IP where i run curl cmd.

**Comment by linsun** [comment]
Just had another discussion with @nrjpoddar on this.  So the XFF header value is working with proxy protocol is enabled.  The only thing that is misleading is ` "X-Envoy-Internal": "true"`, which should be false as it is not providing an envoy internal address.  I suggest we note that as we document the proxy protocol stuff.

**Comment by linsun** [comment]
for the record, another experiment i did on a cluster without proxy protocol or gateway topology.  Note i didn't get my laptop ip as the XFF value, and  `"X-Forwarded-For": "10.188.142.222"` is my gw internal IP. So proxy protocol does work for users to get the correct XFF value.

```
$ curl 169.x/get\?show_env\=true          
{
  "args": {
    "show_env": "true"
  }, 
  "headers": {
    "Accept": "*/*", 
    "Content-Length": "0", 
    "Host": "169.x", 
    "User-Agent": "curl/7.64.1", 
    "X-B3-Parentspanid": "258c2e50c8fbb3c1", 
    "X-B3-Sampled": "0", 
    "X-B3-Spanid": "d921b50b46d2ce23", 
    "X-B3-Traceid": "22fc2bd5fdb11819258c2e50c8fbb3c1", 
    "X-Envoy-Attempt-Count": "1", 
    "X-Envoy-Internal": "true", 
    "X-Forwarded-Client-Cert": "By=spiffe://cluster.local/ns/default/sa/httpbin;Hash=71b3058204e1767176489c16ae06022444c74cd700de826a2114b9e4aff6ed25;Subject=\"\";URI=spiffe://cluster.local/ns/istio2/sa/istio-ingressgateway-service-account", 
    "X-Forwarded-For": "10.188.142.222", 
    "X-Forwarded-Proto": "http", 
    "X-Request-Id": "5c748cd0-d98c-476e-8c08-4ee7c538c134"
  }, 
  "origin": "10.188.142.222", 
  "url": "http://169.x/get?show_env=true"
}
```

File: content/en/docs/ops/configuration/traffic-management/network-topologies/index.md [hunk 7]
```diff
@@ -195,3 +174,33 @@ where `ENUM_VALUE` can be of the following type.
...
+        listener_filters:
+        - name: envoy.listener.proxy_protocol
+        - name: envoy.listener.tls_inspector
+  workloadSelector:
+    labels:
+      istio: ingressgateway
+{{< /text >}}
+
+{{< idea >}}
+Note that the PROXY protocol can be used in conjunction with the `gatewayTopology` configuration, and the `numTrustedProxies` will be used to determine the address for the `X-Envoy-External-Address` header.
```
Comment (nrjpoddar) [inline content/en/docs/ops/configuration/traffic-management/network-topologies/index.md:205]:
```suggestion
Note that the PROXY protocol can be used in conjunction with the `gatewayTopology` configuration, and the `numTrustedProxies` will be used to determine the address for the `X-Forwarded-For` header.
```

**Review by nrjpoddar — COMMENTED** [review]
@linsun found out that Envoy doesn't set `X-Envoy-External-Address` header when PROXY protocol is set but only `X-Forwarded-For` header. cc @PiotrSikora

File: content/en/docs/ops/configuration/traffic-management/network-topologies/index.md (hunk 7 above)
Comment (rvennam) [reply content/en/docs/ops/configuration/traffic-management/network-topologies/index.md:205]: My take away from Lin's tests:

When proxy protocol with gateway listener filter is set and
1. no XFF headers sent, no gatewayTopology = X-Envoy-External-Address not set, XFF set to client IP
2. with XFF headers, no gatewayTopology = X-Envoy-External-Address set, XFF set to client IP
3. with XFF headers, with gatewayTopology = X-Envoy-External-Address set to configured hops, XFF has client IP appended

Without proxy protocol, 
4. no XFF headers sent, no gatewayTopology = X-Envoy-External-Address not set, XFF set to gw IP
5. with XFF headers sent, no gatewayTopology = X-Envoy-External-Address not set, XFF has gw IP appended
6.  no XFF headers, with gatewayTopology = X-Envoy-External-Address not set, XFF set to gw IP
7. with XFF headers, with gatewayTopology = X-Envoy-External-Address set to configured hops, gw IP appended

(...a table might have been a better choice)

This Note was meant to talk about 3.... XFF has client IP appended, but XEEA is set based on numTrustedProxies. 

Agree? We can add another sentence in the description of Proxy Protocol that the XFF value will be set to the client address (or appended).

Comment (nrjpoddar) [reply content/en/docs/ops/configuration/traffic-management/network-topologies/index.md:205]: Mostly. Minor tweaks:

When proxy protocol with gateway listener filter is set and

2. with XFF headers, no gatewayTopology = X-Envoy-External-Address set, XFF set to client IP -> with XFF headers, no gatewayTopology = X-Envoy-External-Address set, XFF has client IP appended

**Review by nrjpoddar — COMMENTED** [review]
@rvennam our users would appreciative if you added a table explaining how proxy protocol, gateway topology, XFF and X-Envoy-External-Address.

Based on @linsun experiments I think that Envoy is not doing the correct thing here but there is another aspect. This excerpt is from Envoy's documentation:

XFF is what Envoy uses to determine whether a request is internal origin or external origin. If use_remote_address is set to true, the request is internal if and only if the request contains no XFF and the immediate downstream node’s connection to Envoy has an internal (RFC1918 or RFC4193) source address. If use_remote_address is false, the request is internal if and only if XFF contains a single RFC1918 or RFC4193 address.

@linsun I would have expected to see the public IP address of your client but I think because a private RFC address (192.168.\*) is set by the client here (curl) directly in the connection string, Envoy is categorizing it as an "internal" request purely based on "192.168.*" client IP address. If the connection string in the PROXY protocol was an external address we would seen a different result i.e. `X-Envoy-External-Address` would have been set.

tl; dr We are seeing weird behavior because you're using `curl --haproxy-protocol` which will append it's private IP address in connection string and is throwing things off with Ingress GW.

Adding @howardjohn @PiotrSikora to get more insights.

Ram, your doc improvements are really helpful, and I hate to complicate this PR but we need to iron this out to give accurate info to our users.

**Comment by rvennam** [comment]
@nrjpoddar , I confirmed your theory about `curl --haproxy-protocol` adding it's private ip in connection string, thereby causing envoy to categorize this as an internet request...X-Envoy-External-Address: true being set, and not seeing the real public IP address.

With PROXY protocol set, I see the following regardless of what my topology setting is.
```
$ curl http://test.rvennam.com/get?show_env=true
{
  "args": {
    "show_env": "true"
  },
  "headers": {
    "Accept": "*/*",
    "Content-Length": "0",
    "Host": "test.rvennam.com",
    "User-Agent": "curl/7.64.1",
    "X-B3-Parentspanid": "85416f88a1a4c0f8",
    "X-B3-Sampled": "0",
    "X-B3-Spanid": "88f67cd78bdf9d7a",
    "X-B3-Traceid": "9ce44e6abdc10ad985416f88a1a4c0f8",
    "X-Envoy-Attempt-Count": "1",
    "X-Envoy-External-Address": "71.71.34.4",
    "X-Forwarded-Client-Cert": "By=spiffe://cluster.local/ns/bookinfo/sa/httpbin;Hash=9d2bb31cf0f51d2ebf2de40ea0ab1596b413ba5161d1a8c044ea45c2d07f4d0e;Subject=\"\";URI=spiffe://cluster.local/ns/custom-gateways/sa/custom-ingressgateway-service-account",
    "X-Forwarded-For": "71.71.34.4",
    "X-Forwarded-Proto": "http",
    "X-Request-Id": "9498f50c-e65f-4eb9-927b-748b9b19e1ac"
  },
  "origin": "71.71.34.4",
  "url": "http://test.rvennam.com/get?show_env=true"
}
```

**Review by nrjpoddar — APPROVED** [review]
Thanks @rvennam for incorporating my many suggestions and being patient with the doc edits.

Kudos to @linsun for all the testing, please sync with Ram as we found out that using curl with `haproxy-protocol` option was having side effects.

**Review by frankbu — APPROVED** [review]
